### DISCUSSION TITLE
How to overwrite source files?

### DISCUSSION THREAD

**Discussion by ghost** [area17/twill#1254] [Q&A] [answered]
I'm trying to adjust some functionality of some Twill source files, and found something that says, if you want to overwrite source files, to copy over Twill files to your own project with the original namespace, add the original files to 'exclude-from-classmap' in composer, and add the copied file to 'files' in composer too (https://spectrum.chat/twill/help/override-settingcontroller-to-store-relationship-data-from-browser-field~457cd4c1-29a8-4bc7-b27f-a39f4e109e4d). 

I'm getting the following error: "Cannot declare class A17\Twill\Http\Controllers\Admin\UserController, because the name is already in use"

Below is my composer.json autoload stuff:
```
    "autoload": {
        "psr-4": {
            "App\\": "app/",
            "Database\\Factories\\": "database/factories/",
            "Database\\Seeders\\": "database/seeders/"
        },
        "files": [
            "app/Models/User.php",
            "app/Models/Enums/UserRole.php",
            "app/Repositories/UserRepository.php",
            "app/Http/Controllers/Admin/LoginController.php",
            "app/Http/Controllers/Admin/UserController.php"
        ],
        "exclude-from-classmap": [
            "vendor/area17/twill/src/Models/User.php",
            "vendor/area17/twill/src/Models/Enums/UserRole.php",
            "vendor/area17/twill/src/Repositories/UserRepository.php",
            "vendor/area17/twill/src/Http/Controllers/Admin/LoginController.php",
            "vendor/area17/twill/src/Http/Controllers/Admin/UserController.php"
        ]
    },
```

Am I missing something?

**Comment by pboivin** [comment, marked as answer]
Hey @aarontilt, the `autoload` config looks fine to me, I'm not sure what could be causing this error.

I recently started exploring overrides through Laravel's container for Twill's internal classes: https://github.com/area17/twill/discussions/1232

Do you think this can work for your use case?

**Reply by ghost** [reply]
I've removed the stuff from composer.json and used the linked solution instead. Seems to work, but now the login controller goes back to the login screen even on successful login.

These are the things I'm trying to override:
```
    public function register()
    {
        $this->app->bind(TwillUser::class, User::class);
        $this->app->bind(TwillUserRole::class, UserRole::class);
        $this->app->bind(TwillUserRepository::class, UserRepository::class);
        $this->app->bind(TwillUserController::class, UserController::class);
        $this->app->bind(TwillLoginController::class, LoginController::class);
    }
```

During debugging, it seems to follow the classes that I want it to, and tries to go to the 'users' route, but then gets redirected back to the login page. I'm assuming there's some middleware strangeness going on?

**Reply by ghost** [reply]
Definitely a problem with rebinding the LoginController. I'm gunna have a mess about and see if I can find anything. Ty for the initial answer!

**Reply by pboivin** [reply]
Ok, interesting! I did have some success overriding LoginController in my tests, trying to add extra processing on login/logout. I'm curious to know what you find on your end!
